inst/private/check_and_convert.m

```diff
@@ -1,63 +1,96 @@
+  persistent builtins
+  persistent sp
+  persistent tuple_1_1
+  persistent tuple_0_0
+  persistent list_or_tuple
+  persistent _sym
+  persistent sym_or_str
+  if isempty(builtins)
+    tuple_1_1 = py.tuple({1, 1});
+    tuple_0_0 = py.tuple({int8(0),int8(0)});
 
+    builtins = pyeval("__builtins__");
+    list_or_tuple = py.tuple({builtins.list, builtins.tuple});
+
+    sp = py.sympy;
+    _sym = py.tuple({sp.Basic, sp.MatrixBase});
+    sym_or_str = py.tuple({sp.Basic, sp.MatrixBase, builtins.str});
+  end
+
+  is_list = py.isinstance(var_pyobj, list_or_tuple);
   if is_list
-    n = pyeval(['len(' var_name ')']);
+    var_pyobj = py.list(var_pyobj);
+    % TODO: This only fails for an empty string, probably can be handled in a better way!
+    n = length(var_pyobj);
   else
     n = 1;
   end
 
   obj = {};
   for i=1:n
     if is_list
-      cur_var = sprintf('%s[%d]', var_name, i-1);
+      cur_pyobj = var_pyobj{i};
     else
-      cur_var = var_name;
+      cur_pyobj = var_pyobj;
+    end
+
+    if(py.isinstance(cur_pyobj, sp.Matrix) && isequal(cur_pyobj.shape, tuple_1_1))
+      %TODO: Probably better if supported via pytave
+      % https://bitbucket.org/mtmiller/pytave/issues/63
+      cur_pyobj = cur_pyobj.__getitem__(tuple_0_0);
     end
 
-    pyexec(strjoin({ ...
-       ['temp = ' cur_var],
-        'if isinstance(temp, sp.Matrix) and temp.shape == (1, 1):',
-       ['    ' cur_var ' = temp[0, 0]']}, newl));
+    if py.isinstance(cur_pyobj, builtins.dict)
+      cur_keys = cur_pyobj.keys();
+      dict_to_struct = true;
+      for j = 1:length(cur_keys)
+        %we can convert only dicts with string &/or sym keys for now as struct
+        %members must be string
+        dict_to_struct = dict_to_struct && py.isinstance(cur_pyobj.keys(){j}, sym_or_str);
+      end
+    else
+      dict_to_struct = false;
+    end
 
-    is_dict_with_sym_keys_curvar = pyeval(['isinstance(', cur_var, ', dict) and len(' cur_var '.keys())>0 and isinstance(' cur_var, '.keys()[0], (sp.Basic, sp.MatrixBase))']);
-    is_list_curvar = pyeval(['isinstance(', cur_var, ', (list, tuple))']);
+    is_list_curvar = py.isinstance(cur_pyobj, list_or_tuple);
 
     if is_list_curvar
-      obj{i} = check_and_convert(cur_var);
-    elseif is_dict_with_sym_keys_curvar
-      %if cur_var is dictionary with symbols as keys then convert it to a struct
-      pyexec(strjoin({'_allKeysStr = []',
-                      '_allValues = []',
-                      ['for key, value in ' cur_var '.iteritems():'],
-                      '    _allKeysStr.append(str(key))',
-                      '    _allValues.append(value)'}, newl));
-      _allKeysStr = pyeval('_allKeysStr');
+      obj{i} = check_and_convert(cur_pyobj);
+    elseif dict_to_struct
+      %if cur_var is dictionary with symbols/strings as keys then convert it to a struct
+      %TODO: will be better after https://bitbucket.org/mtmiller/pytave/issues/62  is resolved
```

Comment:
My laptop isn't working from couole of days, I have given it for repairs. I'll check this as soon as I get my laptop back. I think it should be working now as upstream has struct conversion now.